storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java

```diff
@@ -0,0 +1,177 @@
+    private static final Logger LOG = LoggerFactory.getLogger(GenericResourceAwareSchedulingPriorityStrategy.class);
+
+    @Override
+    protected GrasSimulatedUser getSimulatedUserFor(User u, ISchedulingState cluster) {
+        return new GrasSimulatedUser(u, cluster);
+    }
+
+    @Override
+    public List<TopologyDetails> getOrderedTopologies(ISchedulingState cluster, Map<String, User> userMap) {
+        double cpuAvail = cluster.getClusterTotalCpuResource();
+        double memAvail = cluster.getClusterTotalMemoryResource();
+        Map<String, Double> genericAvail = cluster.getClusterTotalGenericResource();
+
+        List<TopologyDetails> allUserTopologies = new ArrayList<>();
+        List<GrasSimulatedUser> users = new ArrayList<>();
+
+        for (User u : userMap.values()) {
+            users.add(getSimulatedUserFor(u, cluster));
+        }
+
+        while (!users.isEmpty()) {
+            Collections.sort(users, new GrasSimulatedUserComparator(cpuAvail, memAvail, genericAvail));
+            GrasSimulatedUser u = users.get(0);
+            TopologyDetails td = u.getNextHighest();
+            if (td == null) {
+                users.remove(0);
+            } else {
+                double score = u.getScore(cpuAvail, memAvail, genericAvail);
+                td = u.simScheduleNextHighest();
+                LOG.info("GRAS SIM Scheduling {} with score of {}", td.getId(), score);
+                cpuAvail -= td.getTotalRequestedCpu();
+                memAvail -= (td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap());
+                for (Map.Entry<String, Double> entry : td.getTotalRequestedGenericResources().entrySet()) {
+                    String resource = entry.getKey();
+                    Double requestedAmount = entry.getValue();
+                    if (!genericAvail.containsKey(resource)) {
+                        LOG.warn("Resource: {} is not supported in this cluster. Ignoring this request.", resource);
+                    } else {
+                        genericAvail.put(resource, genericAvail.get(resource) - requestedAmount);
+                    }
+                }
+                allUserTopologies.add(td);
+            }
+        }
+        return allUserTopologies;
+    }
+
+    protected static class GrasSimulatedUser extends SimulatedUser {
+
+        // Extend support for Generic Resources in addition to CPU and Memory
+        public final Map<String, Double> guaranteedGenericResources;                // resource name : guaranteed amount
+        private Map<String, Double> assignedGenericResources = new HashMap<>();     // resource name : assigned amount
+
+        public GrasSimulatedUser(User other, ISchedulingState cluster) {
+            super(other, cluster);
+
+            Map<String, Double> guaranteedGenericResources = new HashMap<>();
+            Map<String, Double> availGenericResources = cluster.getClusterTotalGenericResource();   // generic resources that are offered
```

Comment:
Instead of using `getClusterTotalGenericResource` here, probably adding below `getClusterTotalGenericResourceType` to cluster would help avoid calculations..
It would looks something like:
```    @Override
    public Set<String> getClusterGenericResourceTypes() {
        Set<String> resourceTypes =
            supervisors.values().parallelStream().map(sup -> sup.getTotalGenericResources().keySet()).flatMap(Set::stream).collect(Collectors.toSet());
        return resourceTypes;
    }```

Reply:
Added